File: python/ISCAD_Analytical.py
import numpy as np
# from scipy import constants as cons
from functions import *
from plots import *


draw_plots = True # also generate plots, or don't

# load parameters from .json
file = 'ISCAD_parameters.json'
params = loadjson(file)
derived_params(params, file) # add derived values to params
globals().update(params) # update params in current python file 


### STATOR RESISTANCE
Rs_DC = params.rho_s * params.l_stack / (params.h_slot * params.w_slot[0])
print("stator DC resistance = ", ("{:.3f}".format(Rs_DC*1e3)), "mOhm")
zeta = ( (np.pi * cons.mu_0 * params.f) / params.rho_s )**0.5 * params.h_slot
K_R = zeta * (np.sinh(2*zeta) + np.sin(2*zeta)) / (np.cosh(2*zeta) - np.cos(2*zeta)) 
Rs_AC = Rs_DC * K_R
print("stator AC resistance = ", ("{:.3f}".format(Rs_AC*1e3)), "mOhm")

n = 20 # sample count for AC resistance frequency sweep
Rs_ACs = np.zeros((int(params.f/n),2)) # 20 long
for x in range(0,np.shape(Rs_ACs)[0]):
    if x!=0:
        Rs_ACs[x,1] = SlotResistanceAC(x*n, Rs_DC, params)
    else:
        Rs_ACs[x,1] = Rs_DC
    Rs_ACs[x,0] = x*n # frequency

### SELF INDUCTANCE
Ls_s = np.pi/2 * cons.mu_0 * (params.r * params.l_stack / params.ag)
print("self inductance = " , ("{:.3f}".format(Ls_s*1e3)), "mH")

### LEAKAGE INDUCTANCE
# for DC current in rectangular slots
lambda_slot = (params.h_slot / 3*params.w_slot[0]) + (params.h_so / params.w_so)
Ls_sigma_DC = cons.mu_0 * params.l_stack * lambda_slot
print("DC leakage inductance = " ,  ("{:.3f}".format(Ls_sigma_DC*1e6)), "uH")
# AC corrected to fundamental f
K_L = 3 / (2*zeta) * (np.sinh(2*zeta) - np.sin(2*zeta)) / (np.cosh(2*zeta) - np.cos(2*zeta)) 
Ls_sigma_AC = cons.mu_0 * params.l_stack * lambda_slot * K_L
print("AC leakage inductance = " ,  ("{:.3f}".format(Ls_sigma_AC*1e6)), "uH")


### MAIN INDUCTANCE
sum = 0
mutuals = np.zeros(round(params.Qs/2/params.p + 1))
for k in range(1,round(params.Qs/2/params.p + 1)):
    linkage = np.cos( params.p * (k-1) * 2*np.pi / params.Qs)**2
    mutuals[k] = linkage
    sum = sum + linkage
Ls_m = Ls_s * sum
print("main inductance = " ,  ("{:.3f}".format(Ls_m*1e3)), "mH")

print("------")
print("stator slot count: ", params.Qs)
print("pole count: ", params.p)
print("number of slots summed for main inductance L_sm: ", round(params.Qs/2/params.p + 1))
# print("analytical self inductance sum result: ", sum)

### MAIN INDUCTANCEs for range of Qs
Ls_m_arr = np.zeros(Qs)
for slot in range(1,Qs+1): 
    Qs_loop = slot
    m_loop = Qs_loop/p
    sum,Ls_m = 0,0
    for k in range(1,round(m_loop/2)):
        x = np.cos( p * (k-1) * (2*np.pi / Qs_loop) ) ** 2
        sum = sum + x
    Ls_m = Ls_s * sum
    Ls_m_arr[slot-1] = Ls_m
    
if draw_plots == True:
    analytical_plots(params, Ls_s, mutuals, Rs_ACs, Ls_m_arr)


# ### POWER ELECTRONICS

# # parameters [SI]
# # ----------------------USER INPUT-----------------------
# V_DC = 48
# f_SW = 25000
# V_DC_rip_max = 2.6
# Apk = 140 # [30s,330A] [5s,500A] 
# # -------------------------------------------------------

# C_DC_min = Apk / ( params.V_DC_rip_max * 2 * np.pi * params.f_SW )

File: Temporal Analysis/ISCAD_Analytical.m
% ISCAD_Analytical - Analytical calculations for ISCAD motor
% Loads parameters from JSON file into 'mparams' struct and performs analytical calculations

clear; clc;

% Load parameters from JSON file into workspace struct 'mparams'
mparams = load_json_parameters('../ISCAD_parameters.json');

% Constants
mu_0 = 4*pi*1e-7; % Permeability of free space [H/m]
f_r = 0; % Rotor Speed
mparams.p = 2;
Vs_rms = 0.03; % Desired stator voltage input (RMS)

%% RESISTANCES
% Use mparams.variable_name syntax
h_slot = 0.01:0.001:0.025; 
Rs_DC = mparams.rho_s * mparams.l_stack ./ (h_slot * mparams.w_slot(1));
f = linspace(0, 500, 50);
figure
hold on
for j = 1:length(h_slot)
    if f < 1
        K_Rs = 1;
    else
        zeta_Rs = sqrt((pi * mu_0 * f) ./ mparams.rho_s) .* h_slot(j);
        K_Rs = zeta_Rs .* (sinh(2*zeta_Rs) + sin(2*zeta_Rs)) ./ (cosh(2*zeta_Rs) - cos(2*zeta_Rs));
    end
    plot(f, K_Rs)
end
hold off
fprintf('Stator DC resistance = %.2e Ohm\n', Rs_DC);
Rs_AC = Rs_DC .* K_Rs;
figure
plot(f, K_Rs)

%%

% As rotor bars are not defined yet, we will strat by setting the total
% cross sectional area of the bars to be equal to the stator bars to have
% equal reistance.
A_rotorbar = mparams.h_slot * mparams.w_slot(1) * mparams.Qs / mparams.Qr;
Rr_DC = mparams.rho_s * mparams.l_stack / A_rotorbar;
zeta_Rr = sqrt((pi*mu_0*mparams.f)/mparams.rho_s) * (sqrt(4*A_rotorbar/(pi))); % Assumes circular cross section of rotor bars
K_Rr = zeta_Rr .* (sinh(2*zeta_Rr) + sin(2*zeta_Rr)) ./ (cosh(2*zeta_Rr) - cos(2*zeta_Rr));

%% SELF INDUCTANCE
Ls_s = pi/2 * mu_0 * (mparams.r * mparams.l_stack / mparams.ag);
fprintf('Self inductance = %.3f H\n', Ls_s);

%% LEAKAGE INDUCTANCE
% For DC current in rectangular slots
f_sw = linspace(100, 40000, 200);
lambda_slot = (mparams.h_slot / (3*mparams.w_slot(1))) + (mparams.h_so / mparams.w_so);
Ls_sigma_DC = mu_0 * mparams.l_stack * lambda_slot;
zeta_Ls = sqrt(pi*mu_0*f_sw/mparams.rho_s)*mparams.h_slot;
K_Ls = (3./(2*zeta_Ls)).*(sinh(2*zeta_Ls)-sin(2*zeta_Ls))./(cosh(2*zeta_Ls)-cos(2*zeta_Ls));
Ls_sigma_AC = mu_0 * mparams.l_stack * lambda_slot * K_Ls;

fprintf('Leakage inductance DC = %.3f H\n', Ls_sigma_DC);

figure
plot(f_sw, Ls_sigma_AC)


%% MAIN INDUCTANCE
sum_val = 0;
for k = 1:round(mparams.Qs/2/mparams.p)
    sum_val = sum_val + cos(mparams.p * (k-1) * 2*pi / mparams.Qs)^2;
end
Ls_m = Ls_s * sum_val;

fprintf('Main inductance = %.3f H\n', Ls_m);
